t/cobrand/bromley.t

```diff
@@ -1775,6 +1778,83 @@ subtest 'check direct debit reconcilliation' => sub {
     is $failed_new_sub->state, 'hidden', 'failed sub still hidden on second run';
     $ad_hoc_skipped->discard_changes;
     is $ad_hoc_skipped->state, 'unconfirmed', "ad hoc report not confirmed on second run";
+
+};
+
+subtest 'Garden Waste new subs alert update emails contain bin collection days link' => sub {
+    FixMyStreet::override_config {
+        ALLOWED_COBRANDS => 'bromley',
+    }, sub {
+        $mech->clear_emails_ok;
+        my $contact = $mech->create_contact_ok(body => $body, category => 'Garden Subscription', email => 'garden@example.com');
+
+        my $new_sub = ($mech->create_problems_for_body( 1, $body->id, 'Garden Subscription - New', {
+            category => 'Garden Subscription',
+            latitude => 51.402096,
+            longitude => 0.015784,
+            cobrand => 'bromley',
+            cobrand_data => 'waste',
+            areas => '8141',
+            user => $user,
+            send_method_used => 'Open311',
+            whensent => 'now()',
+            external_id => '456',
+            extra => {
+                contributed_by => $staffuser->id,
+                Subscription_Type => 1,
+                Subscription_Details_Quantity => 1,
+                payment_method => 'direct_debit',
+                property_id => '54323',
+                uprn => '654321',
+            },
+        }))[0];
+
+        my $update = FixMyStreet::DB->resultset('Comment')->find_or_create({
+            problem_state => 'action scheduled',
+            problem_id => $new_sub->id,
+            user_id    => $staffuser->id,
+            name       => 'Staff User',
+            mark_fixed => 'f',
+            text       => "Green bin on way",
+            state      => 'confirmed',
+            confirmed  => 'now()',
+            anonymous  => 'f',
+        });
+
+        my $alert = FixMyStreet::DB->resultset('Alert')->create({
+            user => $user,
+            parameter => $new_sub->id,
+            alert_type => 'new_updates',
+            whensubscribed => '2021-09-27 12:00:00',
+            cobrand => 'bromley',
+            cobrand_data => 'waste',
+        });
+        $alert->confirm;
+
+        FixMyStreet::Script::Alerts::send();
+
+        my $alerts = FixMyStreet::App->model('DB::Alert')->search({
+            alert_type => 'new_updates',
+            parameter => $new_sub->id,
+        });
+        is $alerts->count, 1, "alert found for GGW update";
+
+        $mech->email_count_is(1, 'GGW update alert email found');
+        my @emails = $mech->get_email;
+        for my $email (@emails) {
+            my $text_body = $mech->get_text_body_from_email($email);
+            like $text_body, qr/Check your bin collections day/, 'has bin day link text in text part';
+            my @links = $mech->get_link_from_email($email, 'get_all_links');
+            my $found = any { $_ =~ m[recyclingservices\.bromley\.gov\.uk/waste] } @links;
```

Comment:
Why isn't this looking for the property ID in the URL? Doing so would show that I'm afraid that it still isn't working and including the property ID which I assume you expected to be included by the test data. Your test data is what is wrong, property_id is not a key in an extra hash, it is a field. See how `setup_dd_test_report` does it. In fact, why not use `setup_dd_test_report` rather than your own create_problems_for_body call. You could edit it to set cobrand_data = 'waste', then call it as `my $new_sub = setup_dd_test_report({ 'property_id' => '54323' });` and that's all you'd need.

Reply:
I did initially utilise setup_dd_test_report() and am now reverting to do that. I thought extra fields should be handled differently, but then I found some code where a problem was being created with a hashref called 'extra', and borrowed that. I thought this must mean that it would get turned into the extra_fields. It appeared to work (at least in the test), so I didn't look into it any more deeply.

Reply:
Yes, you can also store extra data in a hashref in extra, but that's not how Open311 extra question data (which property_id is treated as) are stored, they're in a _fields list (see Role::Extra for the underlying).